simulator/autoware_auto_msgs_adapter/src/autoware_auto_msgs_adapter_core.cpp

```diff
@@ -0,0 +1,46 @@
+// Copyright 2023 The Autoware Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "autoware_auto_msgs_adapter/autoware_auto_msgs_adapter_core.hpp"
+
+#include "autoware_auto_msgs_adapter/adapter_control.hpp"
+
+#include <rclcpp/rclcpp.hpp>
+
+namespace autoware_auto_msgs_adapter
+{
+
+using autoware_auto_control_msgs::msg::AckermannControlCommand;
+using autoware_control_msgs::msg::Control;
+
+AutowareAutoMsgsAdapterNode::AutowareAutoMsgsAdapterNode(const rclcpp::NodeOptions & node_options)
+: rclcpp::Node("autoware_auto_msgs_adapter", node_options)
+{
+  std::string msg_type_target = declare_parameter<std::string>("msg_type_target");
+  std::string topic_name_source = declare_parameter<std::string>("topic_name_source");
+  std::string topic_name_target = declare_parameter<std::string>("topic_name_target");
+
+  if (msg_type_target == "autoware_auto_control_msgs::msg::AckermannControlCommand") {
+    AdapterControl::SharedPtr adapter =
+      std::make_shared<AdapterControl>(*this, topic_name_source, topic_name_target);
+    adapter_ = std::dynamic_pointer_cast<AdapterBaseInterface>(adapter);
```

Comment:
After asking to ChatGPT "static_pointer_cast vs dynamic_pointer_cast" it said:

> To summarize, `static_pointer_cast` is a compile-time cast that assumes the type conversion is valid, while `dynamic_pointer_cast` is a runtime cast that checks if the conversion is valid and returns a null pointer if it's not. Use `static_pointer_cast` when you're confident about the type relationship and `dynamic_pointer_cast` when you need to perform a safe runtime cast.

I've also checked the documentation for both, I think it's right.

Since:
- My conversion is simple,
- I'm not checking for the conversion result to be null or not (the advantage of `dynamic_pointer_cast`

It's ok to change it with `static_pointer_cast`.

Changed in: https://github.com/autowarefoundation/autoware.universe/pull/3876/commits/f90c0b7ff81e4413a5aa4ca6b41d39e85c6d8085